### PR TITLE
oracle: initial commit

### PR DESCRIPTION
**Pull request by Nosmoht**
##### ISSUE TYPE

New Module Pull Request

##### COMPONENT NAME

Oracle modules

##### ANSIBLE VERSION

```
ansible 2.3.0 (oracle 44e92fa095) last updated 2017/01/31 08:37:28 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

Add modules to support Oracle Database technology. The modules provide support to
- Manage Oracle Database Instance parameters
- Manage Oracle Database tablespaces
- Manage Oracle Database roles
- Manage Oracle Database accounts
- Manage Oracle Database directories

```
```

### CODE REVIEW

**Comment by ryansb** [comment]
For reviews of new modules, it helps significantly to do a PR with a single module (and any needed module_utils code) before submitting all the modules. It both makes the code faster to review, and avoids situations where the same change needs to be made to all modules.

**Comment by ryansb** [comment]
Also, for work-in-progress pull requests, you may choose to title the PR "[WIP] new module xyz to help foo the bars" to let reviewers know the code isn't yet ready.

**Comment by bcoca** [comment]
also please review http://docs.ansible.com/ansible/dev_guide/developing_modules.html

hint: shebang should be `#!/usr/bin/python`, using env breaks interpreter configurations in Ansible.

**Comment by Nosmoht** [comment]
Thanks @ryansb and @bcoca for your comments. I'll follow them next time.

In regards to the interpreter i was quite suprised that the sanity check for module_utils/oracle.py requires 
```
#!/usr/bin/env python
```
whereas all modules required 
```
#!/usr/bin/python
```
Is that expected?

**Comment by bcoca** [comment]
actually that is wrong, module_utils should NOT have shebang (its probably leaking from bin/ansible)

**Comment by Nosmoht** [comment]
Nice to know. But how can i pass the sanity check if i remove shebang?

I got the following result before i changed shebang with https://github.com/ansible/ansible/pull/20848/commits/811f72778b37ab0cdce0a93aabaf0ae39d99ae43
```
2017-01-31 13:42:42 ./lib/ansible/module_utils/oracle.py:#!/usr/bin/python
2017-01-31 13:42:42 One or more file(s) listed above have an unexpected shebang.
2017-01-31 13:42:42 See test/sanity/code-smell/shebang.sh for the list of acceptable values.
2017-01-31 13:42:42 ERROR: Command "test/sanity/code-smell/shebang.sh" returned exit status 1.
```

After the change the check was passed.

**Comment by bcoca** [comment]
check is wrong @mattclay

**Comment by mattclay** [comment]
@Nosmoht Sorry about the confusion on the shebang. The current script verifies that a shebang is on the approved list. However, it doesn't yet indicate when a file should have no shebang. If you remove the shebang, tests will pass.

**Comment by Nosmoht** [comment]
@mattclay Thanks for the hint. Shebang removed and test passed.

_ready_for_review_

**Comment by oswalya** [comment]
Nice module, works really great for my standard setup. So from my point of view we should _shipit_

**Comment by gundalow** [comment]
https://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html

**Comment by gundalow** [comment]
This page should help with the errors in the above comment https://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html

File: lib/ansible/modules/database/oracle/oracle_directory.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    default: present
+    choices: ["present", "absent"]
+version_added: "2.4"
+author: "Thomas Krahn (@nosmoht)"
+extends_documentation_fragment:
+- oracle
+'''
+
+EXAMPLES = '''
+- name: Ensure directory
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_directory.py]: The name seems to be incomplete
`- name: Ensure directory exists`?

File: lib/ansible/utils/module_docs_fragments/oracle.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    default: 1521
+  oracle_user:
+    description:
+    - Account to connect as
+    required: False
+    default: SYSTEM
+  oracle_pass:
+    description:
+    - Password to be used to authenticate.
+    - Can be omitted if environment variable ORACLE_PASS is set.
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/oracle.py]: `C(ORACLE_PASS)`
http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#formatting-options

File: lib/ansible/utils/module_docs_fragments/oracle.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    required: False
+    default: None
+  oracle_mode:
+    description:
+    - Connection mode.
+    - Can be SYSDBA or SYSOPER.
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/oracle.py]: Please remove this line, it repeats whats in `choices:`.

File: lib/ansible/modules/database/oracle/oracle_role.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+- Modify Oracle system parameters
+options:
+  name:
+    description:
+    - Role name
+    required: true
+  roles:
+    description:
+      - List of roles granted to the role.
+      - If an empty list ([]) is specified all granted roles will be revoked.
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_role.py]: `If an empty list C([]) is specified`

File: lib/ansible/modules/database/oracle/oracle_role.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+      - All items will be converted using uppercase.
+    required: false
+    default: None
+  sys_privs:
+    description:
+      - List of system privileges granted to the role.
+      - If an empty list ([]) is specified all granted system privileges will be revoked.
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_role.py]: `If an empty list C([]) is specified`

File: lib/ansible/modules/database/oracle/oracle_role.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    oracle_host: db.example.com
+    oracle_port: 1523
+    oracle_user: system
+    oracle_pass: manager
+    oracle_sid: ORCL
+'''
+
+RETURN = '''
+role:
+  description: Role and attributes as it currently exists within the databsae
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_role.py]: databsae -> database

File: lib/ansible/modules/database/oracle/oracle_system_parameter.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    default: 1521
+  oracle_user:
+    description:
+    - Account to connect as
+    required: False
+    default: SYSTEM
+  oracle_pass:
+    description:
+    - Password to be used to authenticate.
+    - Can be omitted if environment variable ORACLE_PASS is set.
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_system_parameter.py]: `C(ORACLE_PASS)`

File: lib/ansible/modules/database/oracle/oracle_system_parameter.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    required: False
+    default: None
+  oracle_mode:
+    description:
+    - Connection mode.
+    - Can be either of SYSASM, SYSDBA or SYSOPER.
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_system_parameter.py]: Please remove this line.

File: lib/ansible/modules/database/oracle/oracle_tablespace.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    - Amount of datafiles to ensure within the tablespace.
+    - If current amount is lower than specified new datafiles will be added until I(num_datafiles) are reached.
+    - Datafiles will not be removed if more do exist.
+    required: false
+    default: 1
+  autoextend:
+    description:
+    - Boolean to define if autoextend of datafiles is enabled or disabled.
+    required: false
+    default: false
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_tablespace.py:50]: Please add `type: bool` (this is the only `type:` that's supported in DOCUMENTATION.

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#documentation-block

File: lib/ansible/modules/database/oracle/oracle_tablespace.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  init_size:
+    description:
+    - Initial size of a datafile if it will be created.
+    required: false
+    default: 1M
+  next_size:
+    description:
+    - Size of the next extend of each datafile if autoextend is enabled.
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_tablespace.py:60]: Do you need to specify units (MB, GB, etc)?

File: lib/ansible/modules/database/oracle/oracle_tablespace.py [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    required: false
+    default: 0
+  max_size:
+    description:
+    - Max size of each datafile if autoextend is enabled.
+    - Values must by specified as bytes or have one of ['K','M','G','T','P','E'] as suffix.
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_tablespace.py:67]: Does this line need adding to `init_size` and `next_size`?

File: lib/ansible/modules/database/oracle/oracle_user.py [hunk 12]
```diff
@@ -0,0 +1,552 @@
...
+    description:
+    - Password hash as in SYS.USER$.PASSWORD
+    - The I(oracle_user) that connects to the instance must have permissions to read from SYS.USER$.
+  password_mismatch:
+    description:
+    - Boolean to define if a mismatch of current and specified password is allowed.
+    - If I(true), the password will not be changed if different.
+    - If I(false), password will be changed.
+    required: false
+    default: false
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_user.py:56]: please add `type: bool`

File: lib/ansible/modules/database/oracle/oracle_user.py [hunk 13]
```diff
@@ -0,0 +1,552 @@
...
+        return True, client.get_user(name), sql
+    return False, user, sql
+
+
+def main():
+    argument_spec = oracle_argument_spec()
+    argument_spec.update(
+        dict(
+            name=dict(type='str', required=True),
+            password=dict(type='str', required=False),
```
Comment (gundalow) [inline lib/ansible/modules/database/oracle/oracle_user.py]: `, no_log = True`

**Review by gundalow — CHANGES_REQUESTED** [review]
Looks good. Please be aware that in the future it's one module per PR.

Would be good to get some other people to test this and report back.

File: lib/ansible/modules/database/oracle/oracle_user.py (hunk 12 above)
Comment (Nosmoht) [reply lib/ansible/modules/database/oracle/oracle_user.py:56]: Done with https://github.com/ansible/ansible/pull/20848/commits/d0226d5788ebce965ce1b803de529f11d1066d6e

File: lib/ansible/modules/database/oracle/oracle_tablespace.py (hunk 11 above)
Comment (Nosmoht) [reply lib/ansible/modules/database/oracle/oracle_tablespace.py:67]: Done with https://github.com/ansible/ansible/pull/20848/commits/d0226d5788ebce965ce1b803de529f11d1066d6e

File: lib/ansible/modules/database/oracle/oracle_tablespace.py (hunk 10 above)
Comment (Nosmoht) [reply lib/ansible/modules/database/oracle/oracle_tablespace.py:60]: Done with https://github.com/ansible/ansible/pull/20848/commits/d0226d5788ebce965ce1b803de529f11d1066d6e

File: lib/ansible/modules/database/oracle/oracle_tablespace.py (hunk 9 above)
Comment (Nosmoht) [reply lib/ansible/modules/database/oracle/oracle_tablespace.py:50]: Done with https://github.com/ansible/ansible/pull/20848/commits/d0226d5788ebce965ce1b803de529f11d1066d6e

**Comment by Nosmoht** [comment]
@gundalow could you please help me with the following issue? I'm pretty sure i'm not allowed to fix it.
```
test/units/modules/system/test_systemd.py:36:161: E501 line too long (165 > 160 characters)
```

**Comment by mattclay** [comment]
@Nosmoht CI failure was unrelated to this PR. The issue has been resolved and CI is now passing.

**Comment by Nosmoht** [comment]
Hi Guys,

as all checks have been passed which additional changes are required to remove "Changes requested"?

**Comment by gundalow** [comment]
@Nosmoht Add a comment `ready_for_review`

see https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#for-pullrequest-submitters for more info

**Review by gundalow — APPROVED** [review]

**Comment by Nosmoht** [comment]
_ready_for_review_

File: lib/ansible/modules/database/oracle/.gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+.idea
```
Comment (mattclay) [inline lib/ansible/modules/database/oracle/.gitignore:1]: This shouldn't be needed as it already exists in the top-level .gitignore file.

**Comment by Akasurde** [comment]
@Nosmoht Are you still working on this ? Thanks

needs_info

**Comment by Nosmoht** [comment]
I'm currently not working on the modules as the PR is now 2 years old. Sorry.

**Comment by gundalow** [comment]
Given above comment,closing

**Pull request closed without merging**